### PR TITLE
Account Metadata - request authority caching

### PR DESCRIPTION
**Pull request by unpluggedk**
Addresses PR #564

### CODE REVIEW

File: MSAL/src/MSALPublicClientApplication.m [hunk 1]
```diff
@@ -97,6 +97,7 @@ @interface MSALPublicClientApplication()
 
 @property (nonatomic) MSIDDefaultTokenCacheAccessor *tokenCache;
 @property (nonatomic) MSALPublicClientApplicationConfig *internalConfig;
+@property (nonatomic) MSIDAccountMetadataCacheAccessor *accountMetadataCache;
```
Comment (oldalton) [inline MSAL/src/MSALPublicClientApplication.m]: why do we duplicate it here?

File: MSAL/src/MSALPublicClientApplication.m [hunk 2]
```diff
@@ -584,7 +593,8 @@ - (void)acquireTokenWithParameters:(MSALInteractiveTokenParameters *)parameters
     };
     
     MSIDDefaultTokenRequestProvider *tokenRequestProvider = [[MSIDDefaultTokenRequestProvider alloc] initWithOauthFactory:self.msalOauth2Provider.msidOauth2Factory
-                                                                                                          defaultAccessor:_tokenCache
+                                                                                                          defaultAccessor:self.tokenCache
+                                                                                                  accountMetadataAccessor:self.accountMetadataCache
```
Comment (oldalton) [inline MSAL/src/MSALPublicClientApplication.m]: theoretically it doesn't need to be passed through token request provider, although I don't see an issue with it

File: MSAL/src/MSALPublicClientApplication.m [hunk 3]
```diff
@@ -1015,9 +1026,14 @@ - (BOOL)removeAccount:(MSALAccount *)account
         return NO;
     }
     
+    if (![self.accountMetadataCache clearForHomeAccountId:account.homeAccountId.identifier
```
Comment (oldalton) [inline MSAL/src/MSALPublicClientApplication.m]: please use account.identifier instead.

File: MSAL/src/instance/oauth2/aad/MSALAADOauth2Provider.m [hunk 4]
```diff
@@ -98,20 +97,25 @@ - (BOOL)removeAdditionalAccountInfo:(MSALAccount *)account
...
-                                           error:error];
+    if (self.accountMetadataCache)
+    {
+        NSURL *cachedURL = [self.accountMetadataCache getAuthorityURL:requestAuthority.url
+                                                        homeAccountId:account.homeAccountId.identifier
+                                                             clientId:self.clientId
+                                                              context:nil
+                                                                error:error];
+        
+        if (!cachedURL) return requestAuthority;
```
Comment (oldalton) [inline MSAL/src/instance/oauth2/aad/MSALAADOauth2Provider.m]: nit: add a log

File: MSAL/src/instance/oauth2/b2c/MSALB2COauth2Provider.m [hunk 5]
```diff
@@ -62,19 +63,25 @@ - (MSALResult *)resultWithTokenResult:(MSIDTokenResult *)tokenResult
...
-                                           error:error];
+    if (self.accountMetadataCache)
+    {
+        NSURL *cachedURL = [self.accountMetadataCache getAuthorityURL:requestAuthority.url
+                                                        homeAccountId:account.homeAccountId.identifier
+                                                             clientId:self.clientId
+                                                              context:nil
+                                                                error:error];
+        if (!cachedURL) return requestAuthority;
+        return [[MSIDB2CAuthority alloc] initWithURL:cachedURL?:requestAuthority.url
```
Comment (oldalton) [inline MSAL/src/instance/oauth2/b2c/MSALB2COauth2Provider.m]: nit: cachedURL?:requestAuthority.url this will never be executed

**Comment by oldalton** [comment]
Looking good to me. Sovereign clouds issue is still outstanding for the follow up.

File: MSAL/src/MSALPublicClientApplication.m (hunk 1 above)
Comment (unpluggedk) [reply MSAL/src/MSALPublicClientApplication.m]: let's remove (internal) from the +internal header, thus removing duplicate

File: MSAL/src/MSALPublicClientApplication.m (hunk 2 above)
Comment (unpluggedk) [reply MSAL/src/MSALPublicClientApplication.m]: discussed offline, there might be a way to better define controller and tokenRequestProvider, but for now this is ok

File: MSAL/src/MSALPublicClientApplication.m (hunk 3 above)
Comment (unpluggedk) [reply MSAL/src/MSALPublicClientApplication.m]: sure

File: MSAL/src/instance/oauth2/aad/MSALAADOauth2Provider.m (hunk 4 above)
Comment (unpluggedk) [reply MSAL/src/instance/oauth2/aad/MSALAADOauth2Provider.m]: sure

File: MSAL/src/instance/oauth2/b2c/MSALB2COauth2Provider.m (hunk 5 above)
Comment (unpluggedk) [reply MSAL/src/instance/oauth2/b2c/MSALB2COauth2Provider.m]: good catch :p

**Pull request merged**